### PR TITLE
Further reduce the search concurrency overhead.

### PR DESCRIPTION
**Pull request by jpountz**
This iterates on #13546 to further reduce the overhead of search concurrency by caching whether the hit count threshold has been reached: once the threshold has been reached, it cannot get "un-reached" again, so we don't need to pay the cost of `LongAdder#longValue()`.

luceneutil on wikibigall returns the following results with a concurrency of 8:

```
                            TaskQPS baseline      StdDevQPS my_modified_version      StdDev                Pct diff p-value
               HighTermMonthSort     2008.33      (1.8%)     1878.71      (0.9%)   -6.5% (  -9% -   -3%) 0.000
                      AndHighLow     1625.91      (1.9%)     1534.70      (1.7%)   -5.6% (  -8% -   -2%) 0.000
                 CountOrHighHigh      172.89     (19.5%)      164.63     (16.4%)   -4.8% ( -34% -   38%) 0.403
                  CountOrHighMed      339.09     (12.5%)      328.17     (10.0%)   -3.2% ( -22% -   22%) 0.369
                    OrNotHighLow     1399.32      (2.4%)     1360.12      (2.2%)   -2.8% (  -7% -    1%) 0.000
                          IntNRQ      979.38     (12.5%)      972.34     (12.0%)   -0.7% ( -22% -   27%) 0.853
           HighTermDayOfYearSort      753.40      (1.8%)      748.88      (1.3%)   -0.6% (  -3% -    2%) 0.224
                        Wildcard      356.03      (3.0%)      355.18      (3.1%)   -0.2% (  -6% -    6%) 0.806
                         Prefix3      728.14      (3.6%)      726.78      (4.6%)   -0.2% (  -8% -    8%) 0.887
                 CountAndHighMed      382.43      (2.0%)      382.32      (2.6%)   -0.0% (  -4% -    4%) 0.970
                         Respell       56.41      (2.2%)       56.55      (1.8%)    0.3% (  -3% -    4%) 0.691
                          Fuzzy2       96.54      (1.7%)       96.94      (1.3%)    0.4% (  -2% -    3%) 0.390
             And2Terms2StopWords      358.25      (2.8%)      360.56      (2.6%)    0.6% (  -4% -    6%) 0.455
                          Fuzzy1      104.77      (1.6%)      105.67      (1.1%)    0.9% (  -1% -    3%) 0.052
                        PKLookup      296.23      (3.0%)      299.35      (2.4%)    1.1% (  -4% -    6%) 0.221
                     CountPhrase        8.60      (2.3%)        8.70      (2.6%)    1.2% (  -3% -    6%) 0.128
                          Phrase       35.20      (2.7%)       35.70      (4.3%)    1.4% (  -5% -    8%) 0.202
                       CountTerm    11340.80      (4.2%)    11503.91      (4.4%)    1.4% (  -6% -   10%) 0.292
                    AndStopWords       87.39      (4.3%)       88.72      (4.3%)    1.5% (  -6% -   10%) 0.258
                CountAndHighHigh      129.53      (3.2%)      131.56      (2.6%)    1.6% (  -4% -    7%) 0.088
                       And3Terms      372.63      (2.5%)      380.13      (2.3%)    2.0% (  -2% -    6%) 0.008
                     OrStopWords       90.38      (6.0%)       92.38      (5.9%)    2.2% (  -9% -   15%) 0.242
            HighTermTitleBDVSort       67.17      (4.1%)       69.87      (5.5%)    4.0% (  -5% -   14%) 0.009
                      OrHighRare      307.16      (9.2%)      323.41     (10.6%)    5.3% ( -13% -   27%) 0.092
              Or2Terms2StopWords      307.97      (3.0%)      324.57      (2.8%)    5.4% (   0% -   11%) 0.000
                      AndHighMed      446.56      (2.6%)      473.50      (2.8%)    6.0% (   0% -   11%) 0.000
                      TermDTSort      342.04      (2.6%)      363.80      (2.9%)    6.4% (   0% -   12%) 0.000
                     AndHighHigh      196.61      (3.8%)      209.92      (4.1%)    6.8% (  -1% -   15%) 0.000
                        Or3Terms      333.56      (2.7%)      361.92      (3.2%)    8.5% (   2% -   14%) 0.000
                      OrHighHigh      196.12      (5.4%)      213.25      (5.4%)    8.7% (  -1% -   20%) 0.000
                    OrNotHighMed      387.36      (5.5%)      426.70      (5.1%)   10.2% (   0% -   21%) 0.000
                   OrHighNotHigh      291.03      (4.5%)      334.65      (6.4%)   15.0% (   3% -   27%) 0.000
                   OrNotHighHigh      296.55      (4.6%)      347.48      (7.8%)   17.2% (   4% -   31%) 0.000
                       OrHighMed      476.07      (3.0%)      558.12      (3.0%)   17.2% (  10% -   24%) 0.000
               HighTermTitleSort      161.76      (4.2%)      193.64      (2.7%)   19.7% (  12% -   27%) 0.000
                       OrHighLow      811.18      (2.3%)      971.89      (3.8%)   19.8% (  13% -   26%) 0.000
                    OrHighNotMed      399.05      (3.7%)      494.65      (5.1%)   24.0% (  14% -   34%) 0.000
                         LowTerm      940.62      (3.6%)     1196.22      (6.2%)   27.2% (  16% -   38%) 0.000
                         MedTerm      628.29      (3.6%)      806.56      (5.8%)   28.4% (  18% -   39%) 0.000
                    OrHighNotLow      521.71      (4.1%)      671.57      (5.7%)   28.7% (  18% -   40%) 0.000
                        HighTerm      473.33      (3.8%)      613.79      (5.5%)   29.7% (  19% -   40%) 0.000
```

### CODE REVIEW

**Comment by original-brownbear** [comment]
LGTM :)

**Review by easyice — APPROVED** [review]
Nice optimization! 👍

File: lucene/core/src/java/org/apache/lucene/search/HitsThresholdChecker.java [hunk 1]
```diff
@@ -24,6 +24,10 @@ abstract class HitsThresholdChecker {
   /** Implementation of HitsThresholdChecker which allows global hit counting */
   private static class GlobalHitsThresholdChecker extends HitsThresholdChecker {
     private final LongAdder globalHitCount = new LongAdder();
+    // Cache whether the threshold has been reached already. It is not volatile or synchronized on
+    // purpose to contain the overhead of reading the value similarly to what String#hashCode()
+    // does. This does not affect correctness.
+    private boolean thresholdReached = false;
```
Comment (javanna) [inline lucene/core/src/java/org/apache/lucene/search/HitsThresholdChecker.java:30]: I bumped into this only now, and I was wondering: this is mutable state accessed concurrently by multiple threads? If so can it be a primitive boolean without any thread-safety measure?

Comment (javanna) [reply lucene/core/src/java/org/apache/lucene/search/HitsThresholdChecker.java:30]: sounds good, thanks for clarifying

**Pull request merged**